src/program.cpp

```diff
@@ -105,829 +63,148 @@ program::program(const program& p) { assign(p); }
-        {
-            auto&& name = any_cast<builtin::param>(ins.get_operator()).parameter;
-            params.insert(name);
-        }
-    }
-    erase_if(result, [&](auto&& name) { return params.count(name) == 0; });
-    return result;
+    return main_module.get_parameter_names();
 }
 
 instruction_ref program::get_parameter(std::string name) const
 {
-    auto ins = std::find_if(
-        impl->instructions.begin(), impl->instructions.end(), [&](const instruction& x) {
-            if(x.name() == "@param")
-            {
-                return any_cast<builtin::param>(x.get_operator()).parameter == name;
-            }
-            else
-            {
-                return false;
-            }
-        });
-    if(ins != this->end())
-        return ins;
-    else
-        return this->end();
+    return main_module.get_parameter(std::move(name));
 }
 
 std::unordered_map<std::string, shape> program::get_parameter_shapes() const
 {
-    std::unordered_map<std::string, shape> result;
-    for(auto&& ins : impl->instructions)
-    {
-        if(ins.name() == "@param")
-        {
-            auto&& name  = any_cast<builtin::param>(ins.get_operator()).parameter;
-            result[name] = ins.get_shape();
-        }
-    }
-    return result;
+    return main_module.get_parameter_shapes();
 }
 
 bool program::has_instruction(instruction_ref ins) const
 {
-    return std::find_if(
-               impl->instructions.begin(), impl->instructions.end(), [&](const instruction& x) {
-                   return std::addressof(*ins) == std::addressof(x);
-               }) != impl->instructions.end();
+    return main_module.has_instruction(ins);
 }
 
-std::size_t program::size() const { return impl->instructions.size(); }
-instruction_ref program::begin() const { return impl->instructions.begin(); }
-instruction_ref program::end() const { return impl->instructions.end(); }
+std::size_t program::size() const { return main_module.size(); }
+instruction_ref program::begin() const { return main_module.begin(); }
+instruction_ref program::end() const { return main_module.end(); }
 
-std::vector<shape> program::get_output_shapes() const
-{
-    auto last_ins = impl->instructions.back();
-    if(last_ins.name() == "@return")
-    {
-        const auto& output_ins = last_ins.inputs();
-        std::vector<shape> output_shapes;
-        std::transform(output_ins.begin(),
-                       output_ins.end(),
-                       std::back_inserter(output_shapes),
-                       [](auto& ins) { return ins->get_shape(); });
-
-        return output_shapes;
-    }
-    // The else branch is to provide backward compatibility
-    else
-    {
-        return {last_ins.get_shape()};
-    }
-}
+std::vector<shape> program::get_output_shapes() const { return main_module.get_output_shapes(); }
 
-context& program::get_context() const { return impl->ctx; }
+context& program::get_context() const { return main_module.get_context(); }
```

Comment:
The context should be stored in the program and not the module.

Reply:
Looks like each module also need the ctx since it has a finalize() method, in which it calls the ins->finalize(ctx) function. Maybe we can change the module.finalize() to module.finalize(ctx). Is that OK?